File: internal/photoapi/share.go
package photoapi

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/panbotka/kukatko/internal/imgconvert"
	"github.com/panbotka/kukatko/internal/photos"
)

// maxShareFiles caps how many photos one share manifest may describe. It is
// deliberately the ZIP download's cap: both answer the same question — "how much
// may one gesture ask for?" — and a selection over it is told so before anything
// is prepared. The client then hands the files to the phone in batches of its own
// (a share sheet cannot take a thousand files at once), so this cap is the outer
// bound on a whole share, not on one handoff.
const maxShareFiles = maxZipFiles

// shareManifestRequest is the JSON body of the share-manifest endpoint: the
// photos, in the client's selection order, that are about to be handed to the
// phone's share sheet. A UID with no matching photo is silently skipped, as with
// the single download and the ZIP.
type shareManifestRequest struct {
	PhotoUIDs []string `json:"photo_uids"`
}

// shareManifestFile describes one file the client is to fetch and hand over. It
// is deliberately not a URL: the client already knows how to address a photo's
// original and its cached previews (and must be able to retry a preview at a
// smaller size), so the manifest states the *facts* of the file and leaves the
// addressing to the caller.
type shareManifestFile struct {
	// UID identifies the photo whose bytes to fetch.
	UID string `json:"uid"`
	// Name is the file name to hand over, de-duplicated within one manifest so two
	// photos called IMG_0001.jpg do not arrive under one name. A preview carries a
	// .jpg name, because that is what its bytes are.
	Name string `json:"name"`
	// Mime is the type to label the file with — the original's stored type, or
	// image/jpeg for a preview.
	Mime string `json:"mime"`
	// Size is how many bytes to budget for this file when splitting a selection
	// into batches a phone can hold. For a preview it is the *original's* size,
	// which is an upper bound (a JPEG preview is always smaller than the RAW it
	// came from): over-estimating only makes a batch smaller, while
	// under-estimating would let a batch outgrow the phone's memory.
	Size int64 `json:"size"`
	// Preview asks the client to fetch the photo's largest cached JPEG preview
	// instead of the original. Set for a RAW original: a phone library handles a
	// CR2 badly or not at all, and the point of sharing is a picture that lands in
	// Fotky.
	Preview bool `json:"preview"`
}

// shareManifestResponse is the JSON body of the share-manifest endpoint.
type shareManifestResponse struct {
	Files []shareManifestFile `json:"files"`
}

// handleShareManifest answers what a selection looks like as files, so the page
// can fetch them and hand them to the phone's own share sheet (whence iOS offers
// "Save Images" into Apple Photos and Android offers Google Photos).
//
// It exists because the page holds only UIDs — a selection survives scrolling, so
// the photo rows behind it may long have been evicted from the grid's window —
// while the batching it must do needs each file's name, type and size. One request
// answers for the whole selection instead of one detail fetch per photo.
//
// It authorises exactly like the single-photo download (the requireDownload
// guard): naming a file is a step towards fetching it, and the fetch itself goes
// through that same guard. A request naming no resolvable photo is answered with
// 400, one over maxShareFiles with 413, neither having read anything.
func (a *API) handleShareManifest(w http.ResponseWriter, r *http.Request) {
	var req shareManifestRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if len(req.PhotoUIDs) > maxShareFiles {
		writeShareTooMany(w, len(req.PhotoUIDs))
		return
	}
	if len(req.PhotoUIDs) == 0 {
		writeError(w, http.StatusBadRequest, "no photos to share")
		return
	}
	list, err := a.store.ListByUIDs(r.Context(), req.PhotoUIDs)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "collecting photos failed")
		return
	}
	ordered := orderByUIDs(req.PhotoUIDs, list)
	if len(ordered) == 0 {
		writeError(w, http.StatusBadRequest, "no photos to share")
		return
	}
	writeJSON(w, http.StatusOK, shareManifestResponse{Files: shareManifestFiles(ordered)})
}

// shareManifestFiles maps photos to their manifest entries, in order, giving each
// one a name no other entry in the same manifest carries. The names go through the
// ZIP's own sanitising and collision rules — a shared file lands in someone's
// photo library under this name, so it must be a plain base name, and two files
// arriving as one name is exactly the confusion " (2)" prevents.
func shareManifestFiles(list []photos.Photo) []shareManifestFile {
	out := make([]shareManifestFile, 0, len(list))
	used := make(map[string]struct{}, len(list))
	for _, photo := range list {
		preview := imgconvert.IsRAWName(photo.FileName)
		name := sanitizeEntryName(photo.FileName)
		if preview {
			name = jpegFileName(name)
		}
		out = append(out, shareManifestFile{
			UID:     photo.UID,
			Name:    uniqueEntryName(name, used),
			Mime:    shareMime(photo, preview),
			Size:    photo.FileSize,
			Preview: preview,
		})
	}
	return out
}

// shareMime is the type to label a shared file with: image/jpeg for a preview
// (whatever the RAW original's stored type says), the photo's own type otherwise,
// falling back to the generic binary type for a row that never recorded one — a
// File with an empty type is one a share sheet may refuse outright.
func shareMime(photo photos.Photo, preview bool) string {
	if preview {
		return "image/jpeg"
	}
	if photo.FileMime == "" {
		return "application/octet-stream"
	}
	return photo.FileMime
}

// writeShareTooMany answers a selection larger than maxShareFiles with 413 and a
// message naming the cap, so the client can tell the user to share fewer at once.
func writeShareTooMany(w http.ResponseWriter, count int) {
	writeError(w, http.StatusRequestEntityTooLarge,
		fmt.Sprintf("too many photos: %d requested, at most %d per share", count, maxShareFiles))
}

File: web/src/services/share.ts
import { SHARE_PREVIEW_SIZES, type ShareManifestFile } from '../lib/photoShare'

import { ApiError } from './auth'
import { downloadUrl, thumbUrl } from './photos'

/**
 * Client of everything sharing photos out of the library needs from the backend:
 * `POST /api/v1/photos/share-manifest` (what a selection is as files) and the
 * fetches that turn one of those entries into a `File` the phone's share sheet can
 * take.
 *
 * Both fetches ask the media routes to **stream** (`proxy=true`). The page has to
 * read these bytes, and on a deployment with originals in the object store the
 * plain routes answer with a redirect to another origin — which `fetch` follows and
 * then refuses to let the page read. That redirect is the most likely way this
 * feature would silently fail, so the streaming is explicit here rather than
 * assumed.
 */

const API_BASE = '/api/v1'

/** Standard backend error envelope (`internal/photoapi/http.go`). */
interface ErrorBody {
  error?: string
}

/** Extracts the backend error message from a non-OK response, if present. */
async function readErrorMessage(res: Response): Promise<string> {
  try {
    const body = (await res.json()) as ErrorBody
    if (typeof body.error === 'string' && body.error !== '') {
      return body.error
    }
  } catch {
    // Body was empty or not JSON; fall back to the status text below.
  }
  return res.statusText || `request failed: ${res.status}`
}

/** Response body of `POST /api/v1/photos/share-manifest`. */
interface ShareManifestResponse {
  files: ShareManifestFile[]
}

/**
 * Asks what a selection looks like as files, in the order the UIDs were given
 * (`POST /api/v1/photos/share-manifest`).
 *
 * The page holds UIDs, not photos — a selection outlives the grid rows it was made
 * from — so this one request is what supplies the names, types and sizes the
 * batching needs. A UID the catalogue no longer knows is simply absent from the
 * answer.
 *
 * @throws ApiError on a non-OK response, notably 413 when the selection is over
 *   the per-share cap, so the caller can say so specifically.
 */
export async function fetchShareManifest(
  photoUids: string[],
  signal?: AbortSignal,
): Promise<ShareManifestFile[]> {
  const res = await fetch(`${API_BASE}/photos/share-manifest`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    credentials: 'same-origin',
    body: JSON.stringify({ photo_uids: photoUids }),
    signal,
  })
  if (!res.ok) {
    throw new ApiError(res.status, await readErrorMessage(res))
  }
  const body = (await res.json()) as ShareManifestResponse
  return body.files
}

/**
 * The addresses to try for one manifest entry, in order. An original has exactly
 * one; a preview has one per cached size, largest first, because the biggest
 * preview is the one worth having in a phone library and a smaller one is still
 * far better than sharing nothing.
 */
function shareSourceUrls(file: ShareManifestFile): string[] {
  if (!file.preview) {
    return [downloadUrl(file.uid, { original: true, proxy: true })]
  }
  return SHARE_PREVIEW_SIZES.map((size) => thumbUrl(file.uid, size, null, { proxy: true }))
}

/**
 * Fetches one manifest entry's bytes and wraps them in a `File` named and typed as
 * the manifest says — which is what makes the photo arrive in Apple/Google Photos
 * under its own name instead of as an anonymous blob.
 *
 * A preview falls back down the size chain: a size the library never generated (or
 * cannot generate for this RAW) is a failed response, not a reason to give up on
 * the photo. Only when every address fails does the fetch fail, and then the error
 * names nothing but the status — the caller knows which photo it asked for.
 *
 * @throws ApiError when no address answered.
 */
export async function fetchShareFile(file: ShareManifestFile, signal?: AbortSignal): Promise<File> {
  let last: ApiError | undefined
  for (const url of shareSourceUrls(file)) {
    const res = await fetch(url, { credentials: 'same-origin', signal })
    if (!res.ok) {
      last = new ApiError(res.status, await readErrorMessage(res))
      continue
    }
    const blob = await res.blob()
    return new File([blob], file.name, { type: file.mime === '' ? blob.type : file.mime })
  }
  throw last ?? new ApiError(0, 'no source for this photo')
}
